[builtin] adds new builtin for seeking

xhaggi opened the pull request:
This adds a new built-in command `Seek(<seconds>)` to support instant seeking in the current playing media file. @da-anda it should make every small step lover happy ;)

```
<keymap>
  <FullscreenVideo>
    <keyboard>
      <YOURKEY>Seek(-7)</YOURKEY>
    </keyboard>
  <FullscreenVideo>
</keymap>
```

Related discussion at http://forum.kodi.tv/showthread.php?tid=217616&pid=1957016#pid1957016

xhaggi committed:
[builtin] adds new builtin for seeking

FernetMenta commented on xbmc/utils/SeekHandler.cpp:
```diff
@@ -182,6 +182,33 @@ void CSeekHandler::Seek(bool forward, float amount, float duration /* = 0 */, bo
+  m_percent = m_percentPlayTime + percentPerSecond * seconds;
+  g_infoManager.SetSeekStepSize(seconds);
+
+  if (m_percent > 100.0f) m_percent = 100.0f;
```
please don't place multiple statements in a single line. this style is really ugly if you want to set a breakpoint

xhaggi replied:
sure c/p from the other method and not really introduced by me ;) will do a cosmetic commit which change both.

da-anda commented:
nice, thanks. Should we extend JSON-Rpc to also support it? It already provides seeking by percentage, absolute time code and small/bigsteps, but no relative seeking with a custom step size. From a quick look it might be a bit tricky to add it there (requires more advanced evaluation of the input params)

Tolriq commented:
JSON should be able to call the Seek with Input.Action so not a problem for remote makers I guess to use this.

Paxxi commented on xbmc/utils/SeekHandler.cpp:
```diff
@@ -182,6 +182,33 @@ void CSeekHandler::Seek(bool forward, float amount, float duration /* = 0 */, bo
+  if (g_infoManager.GetTotalPlayTime())
+    percentPerSecond = 100.0f / (float)g_infoManager.GetTotalPlayTime();
+
+  m_percent = m_percentPlayTime + percentPerSecond * seconds;
```
why is m_percentPlayTime a member variable and not a local? it's used in two methods but always reset at the start so no real reason to store state for it.

xhaggi replied:
sure, in this method we can use a local. In Seek() we need the state because the method could be called more than once before we perform the seek. The first block is only executed once at the very first call, the second every time.

Paxxi replied:
ahh my mistake, had missed that party, then keep as is :)

Paxxi commented:
What exactly is this trying to solve? looking at the forum thread I get the impression it's to satisfy one users wish to retain old behaviour?

Furthermore, with this being callable from Builtins raises the question of thread safety, there's no protections around state modifications and I'm not sure we know which thread we're called on anymore.

Paxxi commented on xbmc/utils/SeekHandler.cpp:
```diff
@@ -182,6 +182,33 @@ void CSeekHandler::Seek(bool forward, float amount, float duration /* = 0 */, bo
+  m_analogSeek = false;
+  m_seekDelay = 0;
+
+  if (g_infoManager.GetTotalPlayTime())
```
is this meant to check that we have a non-zero playtime or that we're actually playing something? either make explicit what you mean or use an appropriate method.

xhaggi replied:
we need to protect against division by zero.

Paxxi replied:
then make that clear with != 0 or > 0

Paxxi commented:
also please use c++ casts and not old style c casts, see discussion #6791

xhaggi committed:
[cpp] use c++ style casts in CSeekHandler

xhaggi committed:
[cosmetic] put statement of single line if to new line

Montellese commented:
@Tolriq: JSON-RPC doesn't allow calling the builtins directly and obviously it would be better to intergrate this into `Player.Seek` if it is added.

xhaggi commented:
@Paxxi hope i'll have addressed all your comments. I've re-worked the whole method + make it thread-safe.
@FernetMenta added a cosmetic commit which breaks the statements to new line in `Seek()`

Paxxi commented:
On ipad but looks good from what i see @xhaggi

xhaggi commented:
@Montellese if we want to add this to Player.Seek (JSON-RPC) what do you think about if we differ between seconds and percent by adding a %-sign to the percent value?

Tolriq commented:
@xhaggi this completely breaks the API without any backward compatibility.

xhaggi commented:
@Tolriq better we leaven everything as it is and use "10s" or  "-7s" for seek by seconds?

Tolriq commented:
Sounds better for compatibility and easy to use for remote makers.

But not sure about documentation for new comers, this make the value field quite complex in all it's states now.

@Montellese call but why not an string value like "relative" and a new parameter like "amount" ? Easy to follow and document and avoid mixing things when input is a string.

xhaggi commented:
if we introduce new parameters we should use separate for each condition:
* percent
* seconds
* step
* time

Tolriq commented:
Yes that's the idea :) Just need to keep compatibility too.

da-anda commented:
@xhaggi AFAIK you have to add the new JSON-RPC parameter to the schema description as well: https://github.com/xbmc/xbmc/blob/master/xbmc/interfaces/json-rpc/schema/methods.json#L284

xhaggi commented:
@da-anda thanks for the hint, never touched that section ;). If fine by @Montellese I'll add this.

NedScott commented:
@Paxxi Actually, this feature has long since been requested on the forums, long before the current v15 seeking changes. This is one stone that kills many birds.

Montellese committed:
jsonrpc: add explicitly named parameters to Player.Seek and support seeking by a number of seconds

Montellese commented:
Checkout https://github.com/Montellese/xbmc/commit/49907994600809d813afc7a900210bf1d742c623 as a replacement for your last commit. It adds explicitly named properties to the `value` parameter of `Player.Seek` like you mentioned above but also keeps the currently supported values. That way `percentage`, `time` and `step` can be used either as a pure value or as a named property and `seconds` is only supported as a named property.

This is completely runtime untested (only compile tested).

xhaggi commented:
@Montellese great, i'll cherry-pick but couldn't test it by myself. @Tolriq if we provide a test-build could you do some tests?

Tolriq commented:
I can easily test the no regressions and do quick test about new functions yes, and I promise I'll try to find time to build a VM to be able to build again :)

Montellese commented:
@Tolriq: I triggered a test build at http://jenkins.kodi.tv/job/WIN-32/4204/.

xhaggi commented:
Test-Build: http://mirrors.xbmc.org/test-builds/win32/KodiSetup-20150325-3f8a5c1-builtin-seek-support.exe

Tolriq commented:
@xhaggi @Montellese first thing I do tomorrow :)

Paxxi commented:
thanks @NedScott then it seems saner. no objections from me apart from the nitpicking I pointed out :)

Tolriq commented:
@Montellese 

```
03-26 12:10:28.148  13302-32691/org.leetzone.android.yatsewidgetfree E/JsonRpc﹕ JsonRpc.jsonRpcCall@209: No result response : {"error":{"code":-32601,"message":"Method not found."},"id":1,"jsonrpc":"2.0"}
03-26 12:10:28.149  13302-32691/org.leetzone.android.yatsewidgetfree E/JsonRpc﹕ JsonRpc.jsonRpcCall@210: No result command : {"jsonrpc":"2.0","id":1,"method":"Player.Seek","params":{"playerid":1,"value":56.0}}
``` 
Seems not good :)

```
12:08:57 T:8088 WARNING: JSONRPC: Could not parse type "PVR.Details.Channel"
12:08:57 T:8088 WARNING: JSONRPC: Could not parse type "PVR.Details.ChannelGroup.Extended"
12:08:57 T:8088 WARNING: JSONRPC: Could not parse type "GUI.Property.Value"
12:08:57 T:8088 WARNING: JSONRPC: Could not parse type "Setting.Details.SettingList"
12:08:57 T:8088   ERROR: JSONRPC: Unable to parse JSON Schema definition for ""
12:08:57 T:8088   ERROR: JSONRPC: Invalid JSON Schema definition for method ""
```

Sounds like some {} or , problem is json schema definition.

Montellese commented:
@Tolriq: Thanks for the feedback.

@xhaggi: Please pick and squash https://github.com/Montellese/xbmc/commit/114436a3ae6f49195f0e75544bfe363d400bd1d0.

xhaggi commented:
done, triggered a new test-build. http://jenkins.kodi.tv/view/Windows/job/WIN-32/4218/

xhaggi commented:
@Tolriq new test-build is available. would be nice if you could re-test it.

http://mirrors.xbmc.org/test-builds/win32/KodiSetup-20150326-0ce3e89-builtin-seek-support.exe

Tolriq commented:
lready downloaded just needing to find time sorry. Monday at 80% chance or Tuesday

bossanova808 commented:
Didn't test JSON side of it but it's working great from a keymap...

xhaggi commented:
@Tolriq any chance that you can test it this week?

xhaggi commented:
@jjd-uk it would be nice if you could add a small explaination to the [skip steps](http://kodi.wiki/view/Skip_steps) wiki page as advanced configuration for our smallstep back lovers ;)

Tolriq commented:
@xhaggi I had tested regression yesterday and there's things broken (like move to specific time) but had no time to make an investigation on what the problem is :(

Anyway here's the logs from Yatse :

As you can see hours / minutes are logs, seeking by percentage still work, and no time to test the new parameters.

JsonRpc.doRequest@243: Request : {"jsonrpc":"2.0","id":1,"method":"Player.Seek","params":{"playerid":1,"value":{"hours":1,"minutes":30,"seconds":49}}}
JSON TCP message : {"jsonrpc":"2.0","method":"Player.OnSeek","params":{"data":{"item":{"id":7,"type":"movie"},"player":{"playerid":1,"seekoffset":{"hours":0,"milliseconds":0,"minutes":0,"seconds":49},"speed":1,"time":{"hours":0,"milliseconds":103,"minutes":31,"seconds":38}}},"sender":"xbmc"}}

It seems I had an attak on my webserver so no time to do more until I fix it :(

xhaggi commented:
@Tolriq thanks anyway :+1: will take a look at it

xhaggi commented:
@Montellese after a quick look at your changes I don't think this is an newly introduced issue. Would be nice if you could take a look too.

xhaggi commented:
@Montellese ping

Montellese commented:
Sorry I won't be able to look into it until Monday earliest.

xhaggi commented:
okay, let me know what's wrong so i can fix it ;)

xhaggi commented:
@Montellese don't forget to take a look ;)

xhaggi commented:
@Montellese if you don't find time I'll drop your commit so this could be merged for BETA1.

Montellese commented:
@xhaggi: I should find time tomorrow evening. But feel free to drop my commit if you want to merge it ASAP.

xhaggi commented:
i don't want to wait to long because i'm not sure when we close the merge window ;)

Montellese commented:
@xhaggi: Sorry for the delay. Please pick and squash https://github.com/Montellese/xbmc/commit/0d18c4ddfe036faf2e1b7936772d877e30f437b4. There was a small mistake that broke time based seeking (and some bad formatting).

xhaggi commented:
@Montellese thank you for the fix ;)

jenkins build this please

xhaggi commented:
@Tolriq it would be really nice if you could test it again, if not no problem we'll merge it in and fix it later.

xhaggi committed:
[seek] improves instant seek in seek handler

xhaggi commented:
I've added a new commit which improves the "instant" seek support (if you disabled the seek delay).
As the seek is instantly processed we need to drop each call to `Seek()` during this process.

jenkins build this please

da-anda commented:
shouldn't the seek fix be a separate PR? ;)

xhaggi commented:
mhhhh it does not really fix something it's more an improvement and I'm lazy ;)

xhaggi commented:
@da-anda BTW instant seek is used by the seek built-in so it's related to this PR ;)

xhaggi commented:
jenkins build and merge
